modules/s3/src/main/clojure/xtdb/s3.clj

```diff
@@ -48,6 +46,72 @@
                             (catch NoSuchKeyException _
                               (throw (os/obj-missing-exception k))))))))
 
+(def multipart-size-threshold (* 100 1024 1024))
+(def chunk-size (* 10 1024 1024))
+
+(defn single-object-upload
+  [{:keys [^S3AsyncClient client ^S3Configurator configurator bucket prefix]} k ^ByteBuffer buf]
+  (.putObject client
+              (-> (PutObjectRequest/builder)
+                  (.bucket bucket)
+                  (.key (str prefix k))
+                  (->> (.configurePut configurator))
+                  ^PutObjectRequest (.build))
+              (AsyncRequestBody/fromByteBuffer buf)))
+
+(defn multipart-upload
+  [{:keys [^S3AsyncClient client bucket prefix]} k ^ByteBuffer buf]
+  (let [prefixed-key (str prefix k)
+        initiate-request (-> (CreateMultipartUploadRequest/builder)
+                             (.bucket bucket)
+                             (.key prefixed-key)
+                             ^CreateMultipartUploadRequest (.build))]
+    (-> (.createMultipartUpload client initiate-request)
+        (util/then-compose (fn [^CreateMultipartUploadResponse initiate-response]
+                             (let [upload-id (.uploadId initiate-response)
+                                   byte-buffer-size (.limit buf)
+                                   byte-range (range 0 byte-buffer-size chunk-size)
+                                   completed-parts (->> (map-indexed (fn [idx current-pos]
+                                                                      (let [part-number (int (inc idx))
+                                                                            buf-dup (.duplicate buf)
+                                                                            buf-position current-pos
+                                                                            buf-limit (min byte-buffer-size
+                                                                                           (+ current-pos chunk-size))
+                                                                            chunk (-> buf-dup
+                                                                                      (.position ^Integer buf-position)
+                                                                                      (.limit ^Integer buf-limit))
+                                                                            content-length (- buf-limit buf-position)
+                                                                            _ (log/debug (str "Uploading part number: " part-number))
+                                                                            _ (log/debug (str "Buffer chunk: " chunk))
+                                                                        ;; Waits for upload part to finish before starting next one
+                                                                            ^UploadPartResponse upload-part-response (.join
```

Comment:
I think we need to get rid of the .join here to avoid thread starvation. I am not sure exactly how the async IO in the SDK is achieved but it is possible we will choke an IO thread pool or a default forkjoin pool depending on how system is configured. We probably want to do the coordination using thenCompose/thenComposeAsync ensuring our function returns a future. Lets talk this through? I might have it wrong somehow.

Basically we want to chain in the cpu bound work in thenCompose, thenApply blocks etc

I realise this breaks the one-at-a-time sequencing, but we ought to be able to recover that with some future api methods.

Reply:
Like if we block the .putObject thread that is fine. I think I need to sanity check my understanding of CompletableFuture.thenCompose

Reply:
I think thenCompose fns are pushed into the future if it has not completed so and the pool will run it.